rfcs/theme-tokens.md

```diff
@@ -0,0 +1,247 @@
+# RFC: Theme Tokens
+
+<!--
+An RFC can be anything. A question, a suggestion, a plan. The purpose of this template is to give some structure to help folks write successful RFCs. However, don't feel constrained by this template; use your best judgement.
+
+Tips for writing a successful RFC:
+
+- Simple plain words that make your point, fancy words obfuscate
+- Try to stay concise, but don't gloss over important details
+- Try to write a neutral problem statement, not one that motivates your desired solution
+- Remember, "Writing is thinking". It's natural to realize new ideas while writing your proposal
+-->
+
+---
+
+@miroslavstastny
+
+## Summary
+
+This RFC describes types of tokens available in a Theme, proposes theme override possibilities and **introduces on-demand component tokens**.
+
+## Theme Structure
+
+Fluent UI Components are themed using a Theme injected by `FluentProvider`. The `Theme` is an object which contains `tokens` - named variables. Instead of passing theme tokens directly to components, `FluentProvider` injects them as CSS variables into DOM, components then reference the CSS variables in their `makeStyles` calls:
+
+```jsx
+// Theme object
+const theme = {
+  global: {
+    color: {
+      black: '#000' // token value defined for global.color.black
+    }
+  }
+}
+
+// Application root - creates a div, injects --global-color-black: #000 to its styles
+<FluentProvider theme={theme}>
+  <ThemedComponent />
+</FluentProvider>
+
+// Fluent UI Themed Component
+const useStyles = makeStyles([[
+  null,
+  theme => ({ // typed object which matches the theme object structure...
+    color: theme.global.color.black // ...but value is a css variable reference: 'var(--global-color-black)'
+  })
+]]
+
+const ThemedComponent = () => {
+  const cls = useStyles({})
+  return <div className={cls}>ThemedComponent</div>
+}
+```
+
+Currently, there are two different kinds of tokens - **global** and **alias** tokens.
+
+### Global tokens
+
+Global tokens represent named raw values (constants). One name represents the same value no matter whether it is light, dark or high contrast theme.
+
+```js
+const theme = {
+  global: {
+    color: {
+      black: '#000', // global.color.black will always be #000
+    },
+  },
+};
+```
+
+_Currently, global tokens are the same across all the themes with brand color ramp being the only exception._
+
+### Alias tokens
+
+Alias tokens do not represent direct raw values, but are rather mapped to global tokens. The same alias token is mapped to different global tokens in different themes.
+
+```js
+const themeLight = {
+  alias: {
+    color: {
+      brandForeground: brand.shade10,
```

Comment:
Currently, there is no way how developer can match foreground and background token and be sure there is sufficient contrast ratio in all themes.
We plan to add the check to storybook so that at least we can see if the colors can be used together.
I will also share the feedback with designers.

Reply:
> I think it's important to outline why we have alias tokens
Good point, I will add description to storybook.